### DISCUSSION TITLE
Long term maintainability

### DISCUSSION THREAD

**Discussion by ghost** [helix-editor/helix#6356] [General]
I've been using kakoune for a while, and am considering moving to helix. There are a couple of design decisions that the helix devs have made that make it quite compelling. Overall it feels a bit like helix is to kakoune what neovim is to vim.

I've been following the progress on helix for over a year, and there's one thing that I feel a little apprehensive about when it comes to helix. It seems that the helix team is less worried about adding certain features to the helix core. I think that this is great when it comes to LSP and treesitter support. But there are several features that helix bundles, or that are being considered, that I'm rather skeptical about, specifically: integrated window management, an integrated file tree (file pickers even) and an integrated terminal.

- https://github.com/helix-editor/helix/issues/200
- https://github.com/helix-editor/helix/issues/2295
- https://github.com/helix-editor/helix/issues/1976
- https://github.com/helix-editor/helix/issues/312 (relates to the window management)

I can imagine that people want these features, and if you don't like them you don't have to use them of course. But I do think that for the long term health of the project it's important to consider the added complexity and long term maintenance costs for features like these. One thing that kakoune does very well is bundle only what's necessary and make integration of external tools easy. Comparing the two projects it seems like helix is at least on par in terms of LOC, and actually already larger, when compared to kakoune:

- Kakoune: 32.162 LOC (C++, C++ Headers)
- Helix: 47.992 LOC (RS)
- (Counted with https://codetabs.com/count-loc/count-loc-online.html)

With kakoune adding a fuzzy file picker and file browser has been trivial. Combined with a kitty overlay window it integrates seamlessly. I don't need an integrated terminal because I have my terminal for that. I don't need window management because I have my window manager for that (or could even use kitty splits if I wanted to).

I'm not trying to bash helix. The opposite, I think it's a great project. But I know that it's difficult to draw a line for what features to reject and which to include. I'm a little worried that helix will end up trying to do too much, and that the maintainers will end up spending a lot of effort maintaining features that would be better off being outsourced to external tools, or plugins.

I can imagine that for devs that are coming from neovim, or vscode, that having everything integrated in your editor feels right. And that those things should be added to helix if they're missing. I think though that helix would be better off just focusing on what it does well: editing code/text, and remove everything from the core that could be a plugin or an external tool. Thought it would be good to start a discussion on that topic, as I think it will enhance the long term maintainability of the project.

**Comment by kirawi** [comment]
https://github.com/helix-editor/helix/blob/master/docs/vision.md

Generally, commonly used features are intended to be in core. The features you mentioned also have rationale for not being externalized as a plugin:

- File tree: The tree widget can be reused for other UI elements, so the file tree itself is a free bonus.
- Vim tabs: No discussion yet
- Integrated terminal: Important for workflow, https://github.com/helix-editor/helix/issues/5168#issuecomment-1353570305 https://github.com/helix-editor/helix/issues/1976#issuecomment-1316223463
- Client-server: It's just a commonly used feature, making it easier to use window managers.

Helix isn't trying to replace Kakoune, and follows a different philosophy. I personally feel that Helix is quite maintainable, and a large part of it thanks to using Rust which encourages stricter quality control and adds guardrails against common issues. There's also been a lot of effort in maintaining a good contributor experience to get more eyes on the code.

**Reply by ghost** [reply]
Thanks for linking the vision document, I hadn't noticed that yet. This stands out to me:

> Although we want Helix to be productive out-of-the-box, it's not practical or desirable to cram every useful feature and use case into the core editor. The basics should be built-in, but you should be able to extend it with additional functionality as needed. Right now we're thinking Wasm-based plugins.

I suppose that what "the basics" are is subjective. I would say that treesitter and lsp are part of the "the basics" for a code editor. This is in my opinion where helix trumps kakoune. I know helix isn't trying to compete with kakoune, but as a user I'm evaluating them against eachother. By having built-in treesitter and lsp support I feel the integration of these features is a lot better than it is with kakoune. I would say that these features belong to "the basics" because they belong to the domain of editing code.

However, in my opinion, helix isn't as good as kakoune exactly in the areas that are not strictly part of code editing. File exploring: With about 10 lines of code I'm able to integrate lf into kakoune which is a great file explorer. File picker: I can easily integrate fzf with kakoune for all kinds of usecases. Window management: I can manage my kakoune windows with dwm and connect to a single kakoune server. These features end up being a lot nicer and more configurable exactly because kakoune outsources them.

> I personally feel that Helix is quite maintainable, and a large part of it thanks to using Rust which encourages stricter quality control and adds guardrails against common issues. There's also been a lot of effort in maintaining a good contributor experience to get more eyes on the code.

I don't have experience maintaining either, so I can't comment on this. I can imagine that rust is nice to work with and that it protects against certain kinds of common issues. However, the concern I'm raising has more to do with architecture and overall complexity. Any feature has a maintenance cost, no matter how nice the language it's implemented with is to work with.

> and follows a different philosophy.

Yes maybe that's it. Maybe I'm just more of a https://en.wikipedia.org/wiki/Worse_is_better proponent. I suppose there's no wrong or right here, and maybe the Helix approach is just not for me. Arguably yes, all these features have their usecases. Personally though, I feel that trying to do too much can be the downfall of a project. Whether that's the case here, and where the sweet spot lies is up for debate I guess.

As I said, my philosophy is probably just slightly different. Personally I feel the sweet spot is somewhere in between kakoune and the current state of helix. With the great integration for treesitter and lsp, the improvements to the keymap, but without the window management, a built in terminal and file pickers (but with a great plugin system). But then again, this is not my project, so all I can do is raise the topic. If that is not something that the current maintainers feel is true for them then that is also fine.

**Reply by David-Else** [reply]
In another thread on this subject the idea of adding a terminal to Helix was described as ['the definition of bloat'](https://github.com/helix-editor/helix/issues/1976#issuecomment-1257037573). I agree 100%. 

I think the problem stems from people not using readily available superior terminals like Kitty that can communicate through windows and tabs. For example, a couple of features people seem to imagine they need a built in terminal for can be done with ease:

## Running Tests

I have a split Window in Kitty and can run a cargo test using a keyboard shortcut with Helix running beside it. A script that opens a window called `Terminal` runs whenever I open a Rust project:

```
[keys.normal.space]
t = ":sh kitty @ send-text --match 'title:^Terminal' 'clear \\ncargo test \\n'"
```
Pretty much anything could be automated.

## Opening another terminal in your working directory

["I'm very attached to the vscode workflow where I press a hotkey to create a terminal in the directory I'm working in"](https://github.com/helix-editor/helix/issues/1976#issuecomment-1258563531)

Simply add this to your Kitty config:
```
map ctrl+enter          launch --cwd=current
```
 I can find no rational for adding an entire terminal into a bit of software that runs in a terminal, it seems like an **insane** idea.

**Reply by ghost** [reply]
> I think the problem stems from people not using readily available superior terminals like Kitty that can communicate through windows and tabs. For example, a couple of features people seem to imagine they need a built in terminal for can be done with ease:

Yeah. Instead of baking in certain functionality helix could also document how these usecases could be solved with external tools, and allow for that kind of integration. But I suppose that that too is a bit of a philosophical debate. Do you want to pay the maintenance cost of integrating all these things, or do you want to outsource some of that work to your users, asking that they do a little bit more reading and setup work upfront (which gives them more responsibility, but also more freedom)?

Personally I think that for an open source project run by unpaid volunteers it's important to not take on too much. In the beginning there's a lot of velocity, but once that tapers off it can be hard to maintain a project that has taken on a lot of functionality. And yeah, I also feel that for an application that runs in a terminal to integrate a terminal the costs start to outweigh the benefits quite rapidly.

Plus, it's a bit of a vicious cycle. Once your project integrates the functionality that say, vscode integrates, you'll probably attract a lot of users that are used to vscode, who will expect things to work like they did in vscode. The suckless philosophy comes to mind: https://suckless.org/philosophy/. I don't completely agree with this, and it's a little cynical, but the point is not without merit.

But then again, it's a matter of perspective. I don't think the New Jersey vs. MIT style debate will ever be conclusively solved. The difficult thing is to know where to draw the line.

**Reply by kirawi** [reply]
Yes, I agree that it is a matter of perspective. The benefits of a batteries-included approach include: simplified configuration and a consistent workflow across systems (especially on Windows). There are quite a few issues already for the external programs due to user error, such as missing syntax highlighting and clipboard support. On the other hand, the greater surface area may require contributors to make more extensive changes when submitting a PR. Rust can't protect against logic bugs. Ultimately, the line is usually drawn by @archseer.

**Reply by ZioKli** [reply]
I think the idea that Helix should integrate something out of the box can be answered with: if we integrate it will it be substantially better than just using an alternative. I agree that treesitter and LSP support are about as mandatory as it comes, no one is going to be better off without that, but an entire terminal emulator inside Helix which is already running in a terminal seems a bit absurd to me. The terminal alternatives are SO much better than having an integrated terminal, use Wezterm or Alacritty or one of the many other good terminal emulators. In wezterm I literally just keep two tabs open and tab back and forth, ctrl-tab and I'm in my actual terminal. And that is the dumb and lazy way to do it, there are much better methods others have mentioned. To me at least, integrating a terminal will always be worse than just using the actual terminal. It's 1 keyboard shortcut away no matter how you shake it, so why bother with the substantial effort of including it in the built-in feature set? To me this is very clearly in plugin territory.

**Reply by aral** [reply]
As much of a proponent as I am of integrated, seamless experiences, I have to agree with the above regarding an integrated terminal.

Helix is a terminal app. A terminal inside a terminal app doesn’t make sense to me. It’s akin to including an operating system inside of an app. The app runs in the operating system. If you want to use features of the operating system, you use the operating system.

**Reply by archseer** [reply]
Please channel these thoughts into https://github.com/helix-editor/helix/issues/1976 where we've already stated most of this.

**Comment by yatesco** [comment]
My 2c as a relative outsider to helix:
 - I loved vim's simplicity and focus
 - neovim was exciting because (I thought) it promised a more powerful vim
 - neovim now feels like it has somehow lost vim's focus
 - helix is exciting because (I thought) it promises the simplicity of vim but with necessary capability (LSP, which-key, treesitter, fuzzy finding) and a "better" modal language
 - I'm starting to get nervous that helix is now starting to lose that focus on simplicity

Does helix want to be akin to vim/kakoune or VSCode/JetBrains/Emacs? 

*EDIT* - submitted too quickly(!)

My ideal editor:
 - modal editing with huge bonus for elegant multiple selection/cursors
 - quick fix and jump lists (i.e. a persistent and visible list of something that I can navigate/iterate through)
 - LSP integration, but not popping up all over the place (C-x C-o) over autocomplete, every time
 - power, but hidden away, not in your face - show me my code and answer intelligent questions about it (references, implementations etc.), but only when I ask.
 - use splits, sure, but use the OS for terminals, test runners etc. 

I'm genuinely not sure whether my Vim + a few plugins is still the way to go, or maybe kakoune + LSP plugins...

(I'm not arguing technology, I'm only arguing "feelings" and perception. I qualified the worth of this as 2cents so .... ;-))

**Reply by zetashift** [reply]
I totally understand your wants for having an editor that focuses on simplicity and I do think concerns like this deserve a place of discussion.

However things get tricky here, because simplicity is subjective in this context and that makes things very bikeshedding-y. I wonder tho, what makes you feel like helix is now starting to lose a focus on simplicity?

> Does helix want to be akin to vim/kakoune or VSCode/JetBrains/Emacs?

(not a maintainer) I believe helix very much wants to be its own thing. If one extreme is `vi` and the other one is a JetBrains IDE, than you might, for entertaining purposes place, it in the middle, but the way I see it is, helix is a well-rounded editor focused on speed and good defaults, and uses `tree-sitter` and `lsp` to provide language IDE functionality.

I don't know if tree-sitter and LSP (or helix for that matter) will provide the same kind of legacy as vim or emacs. All I know is that I thoroughly enjoy just cracking open helix and getting straight to coding(except for the nightmare that is JS tooling). 

One thing that might scare people is, the Scheme plugin system. Would that complicate things? Maybe, but a plugin system in Helix won't stop people from overengineering their setups (be it in Scheme, Lua or ones 100 `bash` scripts that do all kinds of plugin-y things.
And as a datapoint, I've been using that fork with the scheme plugin for a while now, and after some initial setup, I have 0 desires to play with it. 
Helix is THAT good out of the box.

**Reply by yatesco** [reply]
excellent response, thanks @zetashift. One concern I have is really about LSP. vim's (and TUI editors in general) "working environment" went from being an editor which can be started closed multiple times as needed (just another CLI tool if you will) to now being involved in a longer lived process; you start your LspEE (LSP Enabled Editor), wait for the LSP server to kick in (which can take many seconds on a large Rust project for example), and now you can start your editing session.

That's not necessarily a good or bad thing, but it does mean you want to do more and more in your LspEE to avoid exiting it. Helix is missing a tree view, so now I need to start integrating it with A.N.Other.TUI.File.Explorer, preferably keeping the "Helix session" alive. 

Previously vim (and other TUI editors) had no long lived processes - the longest lived process in vim was you performing the edits you wanted, and there was little value in keeping vim "alive" because it was so quick to start. LspEE have lost that somewhat.

Yes, I'm somewhat conflating "TUI editor" with "quick to start" and "stateless" (definitely the wrong word, but I hope you know what I mean).
I also sincerely apologise for the clumsy and careless reference to JavaEE - I hope I didn't cause anyone any pain!

(apologies, I'm using very clumsy language - very sleep deprived today - I hope it's clear what I mean)

**Reply by zetashift** [reply]
Sorry for the late reply!

> to now being involved in a longer lived process; you start your LspEE (LSP Enabled Editor), wait for the LSP server to kick in (which can take many seconds on a large Rust project for example), and now you can start your editing session.

While I feel LSP could definitely be improved( see this article https://matklad.github.io/2023/10/12/lsp-could-have-been-better.html that explains my concerns a lot better), ultimately it's a matter of tradeoffs.
The fact that I can use Helix for work in a lot of cases is thanks partly to LSP and some other TUI tools that I molded to fit my needs, is something quite wonderous. There was a time, if I wanted to do Scala, it had to be a big IDE that included tons of things and included a whole kitchen-sink. LSP, despite its faults, is something I don't take for granted.

The tradeoffs are in how the user likes their editing experience. Do you want to have `rename symbol` and inlined diagnostics? Then there aren't many options out there besides some program that continuously scans your project for actionable things. Some people don't need that and do it more like `run make` every time they are ready for it and then collect the feedback of the compiler. 
Helix supports both workflows imo, and does that without getting in your way. In neovim and emacs and even vscode I've struggled with configs, unhelpful errors, laggyness if you enable a plugin. And helix gets you there 80% without any real hassle.
For me that's simplicity. For others it might be bloat.

I think the rest of Helix warts(file tree, more integrations etc) will be eventually solved by some old PRs if they ever get merged or by the plugin system.
But if helix brings you concerns right now, then imo, it really depends on what values you find important in your editor and using what editor brings you the most joy right now? There are no wrong answers when neovim/emacs/helix/kakoune all got tree-sitter and LSP support in some way.
Discomfort with LSP can be overcome or you can decide to not use it, but I don't think you can have refactoring on (big) projects without having some state. LSP can be the enemy you accept or not, but in my experience Helix makes things very easy for me compared to other editors.


(sorry for the braindump :P)

**Reply by yatesco** [reply]
thanks @zetashift  - I agree with everything you say, and I think LSPs being a persistent server that editors connect to (rather than being a process managed and scoped by) would go a *long* way to resolving this.

**Reply by zetashift** [reply]
@yatesco Another option would be if helix could have a daemon mode: https://github.com/helix-editor/helix/issues/312. That way if the 'server' instance of helix holds onto the connection with `rust-analyzer` for example, it won't have to rescan the whole project again.

**Reply by yatesco** [reply]
I'm not sure why LSP servers are designed to be persistent.

**Reply by Ericson2314** [reply]
I only just started trying out Helix, but I do think https://github.com/helix-editor/helix/issues/312 is important an important thing to do even if one is happy with a single terminal window.

**Reply by lukeflo** [reply]
Reviving an old discussion just to add my 2 cent...

> * I loved vim's simplicity and focus
> * neovim was exciting because (I thought) it promised a more powerful vim
> * neovim now feels like it has somehow lost vim's focus
> * helix is exciting because (I thought) it promises the simplicity of vim but with necessary capability (LSP, which-key, treesitter, fuzzy finding) and a "better" modal language
> * I'm starting to get nervous that helix is now starting to lose that focus on simplicity

But that's exactly how I fell. I loved Helix from the start mainly because of its very good default set of features and the simple config.

Neovim with Lazy seemed much too bloated with its multiple Lua config files sourced by its main `init.lua`.

Now having a deeper look at the upcoming `scheme` config/plugin system I'm not sure if I like it. Yes, it makes it more customizable, but the snippets of a new config posted here trigger memories of my Emacs `elisp` config. You had to adjust numerous things to get it working in an acceptable way.

I know a plugin system is wanted by **many** users and I understand the maintainers can't take every individual concern (like this one) into account; they're already doing a great job.

But I hope Helix wont loose its focus on simplicity (very subjective, but let's say in comparison to nvim or Emacs config/package system). As one of those "bash script" guys, I appreciate very much the possibility to have a ready to go in-terminal editor which I can enhance with some scripting stuff, but without running into the rabbit hole of a blown up config file containing thousands of lines and the need to install multiple plugins/packages (whatever you call it, its the same to me) just to set up the editor for some coding/writing...

But I guess you maintaing guys have that in mind.

Keep up the great work!
